File: scenario_mpc/mode_weights.py
"""
Mode weight computation for adaptive scenario-based MPC.

Implements Section 4 of guide.md: Mode History and Weights

Supports three weight computation strategies:
- Uniform: Equal weights for all modes (Eq. 4)
- Recency: Exponential decay weighting recent observations (Eq. 5)
- Frequency: Weights based on observation frequency (Eq. 6)
"""

import numpy as np
from typing import Dict, List, Tuple
from .types import ModeHistory, WeightType


def compute_mode_weights(
    mode_history: ModeHistory,
    weight_type: WeightType = WeightType.FREQUENCY,
    recency_decay: float = 0.9,
    current_timestep: int = 0
) -> Dict[str, float]:
    """
    Compute mode weights based on observation history.

    Following Section 4 of guide.md.

    Args:
        mode_history: Observed mode history for an obstacle
        weight_type: Weight computation strategy
        recency_decay: Decay factor lambda for recency weighting
        current_timestep: Current timestep for recency computation

    Returns:
        Dictionary mapping mode_id to weight (normalized to sum to 1)
    """
    modes = list(mode_history.available_modes.keys())
    num_modes = len(modes)

    if num_modes == 0:
        return {}

    if weight_type == WeightType.UNIFORM:
        # Eq. 4: w_m = 1/M for all modes
        weights = {mode_id: 1.0 / num_modes for mode_id in modes}

    elif weight_type == WeightType.RECENCY:
        # Eq. 5: w_m = sum_{t: m_t = m} lambda^(T - t)
        weights = _compute_recency_weights(
            mode_history, modes, recency_decay, current_timestep
        )

    elif weight_type == WeightType.FREQUENCY:
        # Eq. 6: w_m = n_m / sum_j n_j
        weights = _compute_frequency_weights(mode_history, modes)

    else:
        raise ValueError(f"Unknown weight type: {weight_type}")

    # Normalize weights to sum to 1
    total = sum(weights.values())
    if total > 0:
        weights = {k: v / total for k, v in weights.items()}
    else:
        # Fallback to uniform if no observations
        weights = {mode_id: 1.0 / num_modes for mode_id in modes}

    return weights


def _compute_recency_weights(
    mode_history: ModeHistory,
    modes: List[str],
    decay: float,
    current_timestep: int
) -> Dict[str, float]:
    """
    Compute recency-based weights (Eq. 5).

    w_m = sum_{t: m_t = m} lambda^(T - t)

    Recent observations are weighted more heavily.
    """
    weights = {mode_id: 0.0 for mode_id in modes}

    for timestep, mode_id in mode_history.observed_modes:
        if mode_id in weights:
            # Exponential decay based on how old the observation is
            age = current_timestep - timestep
            weights[mode_id] += decay ** age

    return weights


def _compute_frequency_weights(
    mode_history: ModeHistory,
    modes: List[str]
) -> Dict[str, float]:
    """
    Compute frequency-based weights (Eq. 6).

    w_m = n_m / sum_j n_j

    where n_m is the number of times mode m was observed.
    """
    counts = mode_history.get_mode_counts()
    weights = {mode_id: float(counts.get(mode_id, 0)) for mode_id in modes}
    return weights


def sample_mode_sequence(
    mode_weights: Dict[str, float],
    horizon: int,
    rng: np.random.Generator = None
) -> List[str]:
    """
    Sample a mode sequence for the prediction horizon.

    Assumes modes are i.i.d. across timesteps (can be extended for Markov).

    Args:
        mode_weights: Weights for each mode
        horizon: Number of timesteps to sample
        rng: Random number generator

    Returns:
        List of mode_ids of length horizon
    """
    if rng is None:
        rng = np.random.default_rng()

    modes = list(mode_weights.keys())
    weights = np.array([mode_weights[m] for m in modes])

    # Normalize weights
    weights = weights / weights.sum()

    # Sample mode indices
    indices = rng.choice(len(modes), size=horizon, p=weights)

    return [modes[i] for i in indices]


def sample_mode_from_weights(
    mode_weights: Dict[str, float],
    rng: np.random.Generator = None
) -> str:
    """
    Sample a single mode from the weight distribution.

    Args:
        mode_weights: Weights for each mode
        rng: Random number generator

    Returns:
        Sampled mode_id
    """
    if rng is None:
        rng = np.random.default_rng()

    modes = list(mode_weights.keys())
    weights = np.array([mode_weights[m] for m in modes])
    weights = weights / weights.sum()

    idx = rng.choice(len(modes), p=weights)
    return modes[idx]


def compute_mode_transition_matrix(
    mode_history: ModeHistory,
    modes: List[str]
) -> np.ndarray:
    """
    Estimate mode transition probabilities from history.

    P[i,j] = P(mode_j | mode_i) estimated from observations.

    Args:
        mode_history: Observation history
        modes: List of mode_ids

    Returns:
        Transition matrix (num_modes x num_modes)
    """
    num_modes = len(modes)
    mode_to_idx = {m: i for i, m in enumerate(modes)}

    # Count transitions
    counts = np.zeros((num_modes, num_modes))
    observations = mode_history.observed_modes

    for i in range(len(observations) - 1):
        _, mode_from = observations[i]
        _, mode_to = observations[i + 1]

        if mode_from in mode_to_idx and mode_to in mode_to_idx:
            idx_from = mode_to_idx[mode_from]
            idx_to = mode_to_idx[mode_to]
            counts[idx_from, idx_to] += 1

    # Normalize rows to get probabilities
    row_sums = counts.sum(axis=1, keepdims=True)
    row_sums[row_sums == 0] = 1  # Avoid division by zero

    transition_matrix = counts / row_sums

    # For rows with no observations, use uniform distribution
    zero_rows = (counts.sum(axis=1) == 0)
    transition_matrix[zero_rows, :] = 1.0 / num_modes

    return transition_matrix

File: cpp_mpc/src/mode_weights.cpp
/**
 * @file mode_weights.cpp
 * @brief Implementation of mode weight computation.
 */

#include "mode_weights.hpp"
#include <cmath>
#include <algorithm>

namespace scenario_mpc {

namespace {

/**
 * @brief Compute recency-based weights (Eq. 5).
 *
 * w_m = sum_{t: m_t = m} lambda^(T - t)
 * Recent observations are weighted more heavily.
 */
std::map<std::string, double> compute_recency_weights(
    const ModeHistory& mode_history,
    const std::vector<std::string>& modes,
    double decay,
    int current_timestep
) {
    std::map<std::string, double> weights;
    for (const auto& mode_id : modes) {
        weights[mode_id] = 0.0;
    }

    for (const auto& [timestep, mode_id] : mode_history.observed_modes) {
        if (weights.find(mode_id) != weights.end()) {
            // Exponential decay based on how old the observation is
            int age = current_timestep - timestep;
            weights[mode_id] += std::pow(decay, age);
        }
    }

    return weights;
}

/**
 * @brief Compute frequency-based weights (Eq. 6).
 *
 * w_m = n_m / sum_j n_j
 * where n_m is the number of times mode m was observed.
 */
std::map<std::string, double> compute_frequency_weights(
    const ModeHistory& mode_history,
    const std::vector<std::string>& modes
) {
    auto counts = mode_history.get_mode_counts();
    std::map<std::string, double> weights;

    for (const auto& mode_id : modes) {
        weights[mode_id] = static_cast<double>(counts[mode_id]);
    }

    return weights;
}

}  // anonymous namespace

std::map<std::string, double> compute_mode_weights(
    const ModeHistory& mode_history,
    WeightType weight_type,
    double recency_decay,
    int current_timestep
) {
    std::vector<std::string> modes;
    for (const auto& [mode_id, _] : mode_history.available_modes) {
        modes.push_back(mode_id);
    }

    int num_modes = static_cast<int>(modes.size());
    if (num_modes == 0) {
        return {};
    }

    std::map<std::string, double> weights;

    switch (weight_type) {
        case WeightType::UNIFORM:
            // Eq. 4: w_m = 1/M for all modes
            for (const auto& mode_id : modes) {
                weights[mode_id] = 1.0 / num_modes;
            }
            break;

        case WeightType::RECENCY:
            // Eq. 5: w_m = sum_{t: m_t = m} lambda^(T - t)
            weights = compute_recency_weights(
                mode_history, modes, recency_decay, current_timestep
            );
            break;

        case WeightType::FREQUENCY:
            // Eq. 6: w_m = n_m / sum_j n_j
            weights = compute_frequency_weights(mode_history, modes);
            break;
    }

    // Normalize weights to sum to 1
    double total = 0.0;
    for (const auto& [_, w] : weights) {
        total += w;
    }

    if (total > 0) {
        for (auto& [_, w] : weights) {
            w /= total;
        }
    } else {
        // Fallback to uniform if no observations
        for (auto& [_, w] : weights) {
            w = 1.0 / num_modes;
        }
    }

    return weights;
}

std::vector<std::string> sample_mode_sequence(
    const std::map<std::string, double>& mode_weights,
    int horizon,
    std::mt19937& rng
) {
    std::vector<std::string> sequence;
    sequence.reserve(horizon);

    for (int k = 0; k < horizon; ++k) {
        sequence.push_back(sample_mode_from_weights(mode_weights, rng));
    }

    return sequence;
}

std::string sample_mode_from_weights(
    const std::map<std::string, double>& mode_weights,
    std::mt19937& rng
) {
    std::vector<std::string> modes;
    std::vector<double> weights;

    for (const auto& [mode_id, w] : mode_weights) {
        modes.push_back(mode_id);
        weights.push_back(w);
    }

    // Normalize weights
    double total = 0.0;
    for (double w : weights) {
        total += w;
    }
    for (double& w : weights) {
        w /= total;
    }

    // Sample using discrete distribution
    std::discrete_distribution<int> dist(weights.begin(), weights.end());
    int idx = dist(rng);

    return modes[idx];
}

Eigen::MatrixXd compute_mode_transition_matrix(
    const ModeHistory& mode_history,
    const std::vector<std::string>& modes
) {
    int num_modes = static_cast<int>(modes.size());
    std::map<std::string, int> mode_to_idx;
    for (int i = 0; i < num_modes; ++i) {
        mode_to_idx[modes[i]] = i;
    }

    // Count transitions
    Eigen::MatrixXd counts = Eigen::MatrixXd::Zero(num_modes, num_modes);
    const auto& observations = mode_history.observed_modes;

    for (size_t i = 0; i + 1 < observations.size(); ++i) {
        const std::string& mode_from = observations[i].second;
        const std::string& mode_to = observations[i + 1].second;

        if (mode_to_idx.find(mode_from) != mode_to_idx.end() &&
            mode_to_idx.find(mode_to) != mode_to_idx.end()) {
            int idx_from = mode_to_idx[mode_from];
            int idx_to = mode_to_idx[mode_to];
            counts(idx_from, idx_to) += 1.0;
        }
    }

    // Normalize rows to get probabilities
    Eigen::MatrixXd transition_matrix(num_modes, num_modes);

    for (int i = 0; i < num_modes; ++i) {
        double row_sum = counts.row(i).sum();
        if (row_sum > 0) {
            transition_matrix.row(i) = counts.row(i) / row_sum;
        } else {
            // For rows with no observations, use uniform distribution
            transition_matrix.row(i).setConstant(1.0 / num_modes);
        }
    }

    return transition_matrix;
}

}  // namespace scenario_mpc
